### ISSUE TITLE
FR: Support per-campaign WML unit tests

### ISSUE THREAD

**Issue by stevecotton** [wesnoth/wesnoth#5273]
Feature request: unit tests for campaigns' WML

Illustrative example: SotA has macros `SAVE_ZOMBIE_LIST` and `LOAD_ZOMBIE_LIST`. Bug #5256 could have easily been caught by a unit test, because the root cause is that `{SAVE_ZOMBIE_LIST variable_name}` ignores the `variable_name` and thus overwrites the wrong variable.

I'd like to be able to have a unit-test file that's loaded with the appropriate campaign's WML available, so that it can unit test stuff in the campaign's utils folder.

**Comment by Pentarctagon** [comment]
Campaign WML can already be made available by starting wesnoth while providing the particular campaign define, I think?

**Comment by soliton-** [comment]
Issues like that could also trigger a warning in the preprocessor. A macro with an unused parameter seems suspicious to begin with. Not sure if it even could be an error or there are some legitimate use cases.

**Comment by stevecotton** [comment]
About to open a PR adding this (posting it here because I'm still checking it compiles before pushing it):

```
	if(cmdline_opts.preproc_define || cmdline_opts.preprocess_input_macros || cmdline_opts.preprocess_path) {
		// It would be good if this was supported for running tests too, possibly for other uses.
		// For the moment show an error message instead of leaving the user wondering why it doesn't work.
		std::cerr << "That --preprocess-* option is only supported when using --preprocess or --validate-wml.\n";
		return -1;
	}
```

**Comment by soliton-** [comment]
`run_wml_tests --additional_arg=--campaign=my_campaign --additional_arg=--campaign-difficulty=NORMAL --additional_arg=--campaign-scenario=my_campaign_unit_test`

Does that do anything useful? (Given an actual campaign and scenario ID.)

**Comment by stevecotton** [comment]
@soliton- in do_gameloop(), the play_test() and unit_test() paths run and quit before we reach the call to goto_campaign(). Those options currently won't affect anything run via the unit test support.

**Comment by stevecotton** [comment]
The unit-test framework also seems to run without the terrain macros enabled. I tried dropping `#define CAMPAIGN_SECRETS_OF_THE_ANCIENTS` into the campaign's _main.cfg, but running with `--unit-test` gots errors about `NEW:FENCE` not being defined.

**Comment by stevecotton** [comment]
Clarification of what's blocking this: it's mainly about preprocessing, rather than campaigns.

When running Wesnoth normally, enabling or disabling per-campaign functionality is largely done via the preprocessor. It's just taking a list of strings from the [campaign] tag, possibly some strings from a [scenario] tag and then using the result of preprocessing with those #define'd. It's easy to just copy-and-paste those lists of strings into a test, so a useful implementation could skip all handling of the [campaign] tag.

Configuring the list of strings passed to the preprocessor requires either a hack such as just dumping a bunch of #defines in a .cfg file (see c144e31's hack of data/_main.cfg), or it requires a major refactor of the C++ code around option parsing, wesnoth.cpp, game_launcher.cpp, and a load of global variables that are hidden in the de-facto singleton instances of saved_game and game_config.

On 4 of these 5 there's a different block of code that runs the preprocessor:
* anything that obeys the `--preprocess-defines` argument
* running the WML unit tests
* running the Boost unit tests
* running the GUI2 Boost unit tests for the loading screen
* playing a game
... and I'd prefer a solution that didn't involve writing a 5th version of this code to cover the 5th of those paths...

**Comment by ProditorMagnus** [comment]
Also for addons, not just campaigns.

**Comment by CelticMinstrel** [comment]
I don't think this requires any extra coding. Just add a line in the campaigns _main.cfg, outside the campaign define, which imports any tests in the campaign. I think that should work both for addons and for mainline campaigns.
